### ISSUE TITLE
Request for feedback

### ISSUE THREAD

**Issue by TomerAberbach** [anthropics/anthropic-sdk-java#30]
Hey everyone!

This SDK is in alpha, and we are actively iterating on its design and making improvements. We would love to hear your thoughts and feedback!

As you try the new SDK, please post your thoughts and feedback on this issue. We'll be actively monitoring this thread.

Thanks in advance! We hope to make a Java SDK you'll love using.

**Comment by sgflt** [comment]
**Batch API**

I am stuck at retrieving batch results.

```java
  public MessageBatch getResults(final String batchId) {
    return this.client.messages()
        .batches()
        .retrieve(MessageBatchRetrieveParams.builder().messageBatchId(batchId).build());
  }
```

`MessageBatch` contains `resultsUrl`, but  it seems there is a part of API missing.

I would expect a stream/list of jsonl records to be available in similar way as in python
```python
client.messages.batches.results(
    "msgbatch_01HkcTjaV5uDC8jWR4ZsDV8d",
):
```

[doc](https://docs.anthropic.com/en/api/retrieving-message-batch-results)


**API model**
I feel kind of confused by multiple implementations with the same meaning and structure. F. e.:
```java

var SYSTEM_PROMPT = "system prompt as string";
MessageBatchCreateParams.Request.Params.builder()                            
      .system(MessageBatchCreateParams.Request.Params.System.ofString(SYSTEM_PROMPT))
      // there is no string overload


MessageCreateParams.builder()
    .system(MessageCreateParams.System.ofString(SYSTEM_PROMPT))
    // or
    .system(SYSTEM_PROMPT)
```

**Comment by katoquro** [comment]
Hello,

Is it possible to have a separate library containing only the model types, without additional dependencies? We do not use OkHttp and have our own client, so we would like to have the "official" types and validators available separately. However, having a ready-made client as a separate artifact is beneficial for building smaller projects, like a Lambda function, where minimizing dependencies is important.
Thank you.

**Comment by TomerAberbach** [comment]
> **Batch API**
> 
> I am stuck at retrieving batch results.
> 
>   public MessageBatch getResults(final String batchId) {
>     return this.client.messages()
>         .batches()
>         .retrieve(MessageBatchRetrieveParams.builder().messageBatchId(batchId).build());
>   }
> `MessageBatch` contains `resultsUrl`, but it seems there is a part of API missing.
> 
> I would expect a stream/list of jsonl records to be available in similar way as in python
> 
> client.messages.batches.results(
>     "msgbatch_01HkcTjaV5uDC8jWR4ZsDV8d",
> ):
> [doc](https://docs.anthropic.com/en/api/retrieving-message-batch-results)

Yeah, support for this API is incomplete. JSONL support isn't built out properly yet. Filed an issue here for you to follow: https://github.com/anthropics/anthropic-sdk-java/issues/42

> **API model** I feel kind of confused by multiple implementations with the same meaning and structure. F. e.:
> 
> var SYSTEM_PROMPT = "system prompt as string";
> MessageBatchCreateParams.Request.Params.builder()                            
>       .system(MessageBatchCreateParams.Request.Params.System.ofString(SYSTEM_PROMPT))
>       // there is no string overload
> 
> 
> MessageCreateParams.builder()
>     .system(MessageCreateParams.System.ofString(SYSTEM_PROMPT))
>     // or
>     .system(SYSTEM_PROMPT)

Yup, this is a known issue. We will soon have these convenience overloads recursively :)

**Comment by TomerAberbach** [comment]
> Hello,
> 
> Is it possible to have a separate library containing only the model types, without additional dependencies? We do not use OkHttp and have our own client, so we would like to have the "official" types and validators available separately. However, having a ready-made client as a separate artifact is beneficial for building smaller projects, like a Lambda function, where minimizing dependencies is important. Thank you.

So, the SDK is already split into multiple packages:
- `anthropic-java-client-okhttp` (depends on OkHttp)
- `anthropic-java-core` (doesn't depend on OkHttp)
- `anthropic-java` (depends on both `anthropic-java-client-okhttp` and `anthropic-java-core`)

You should be able to depend on `anthropic-java-core` only and bring your own HTTP client by implementing [`HttpClient`](https://github.com/anthropics/anthropic-sdk-java/blob/main/anthropic-java-core/src/main/kotlin/com/anthropic/core/http/HttpClient.kt) using your client, and then using it to construct the Anthropic clients ([see OkHttp examples](https://github.com/anthropics/anthropic-sdk-java/tree/main/anthropic-java-client-okhttp/src/main/kotlin/com/anthropic/client/okhttp)).

Does that answer your question @katoquro?

**Comment by katoquro** [comment]
Thanks for response, @TomerAberbach. 

I saw that module. Isn't `org.apache.httpcomponents.client5:httpclient5` an http client?  
It's listed here https://github.com/anthropics/anthropic-sdk-java/blob/04dddd52df5569219faca1e0ae31aa968c63fdc7/anthropic-java-core/build.gradle.kts#L16

**Comment by TomerAberbach** [comment]
@katoquro apologies for the delayed reply.

I'll need to look into why we depend on that, but we definitely don't use that as our HTTP client in the core module. Maybe we depend on it for something else, or maybe it's a bug. I'll get back to you on that

**Comment by Mubai0628** [comment]
Sorry to bother you, why do I always get 403 when I try to access Claude through the api in Java? But I succeed directly in python? Why does Java show that ANTHROPIC_AUTH_TOKEN is still needed?

**Comment by sgflt** [comment]
I am trying to implement custom tool, but I find it difficult to find a correct way, how to build shema.

```java
            // TODO custom tool
            .addTool(
                Tool.builder().inputSchema(
                    Tool.InputSchema.builder()
                        .putAdditionalProperty("my_tool_property", /* ??? */)
                        .build()
                )
            )
```


I would expect direct input:
```java
             .addTool(
                 """
                 {
        "name": "get_weather",
        "description": "Get the current weather in a given location",
        "input_schema": {
            "type": "object",
            "properties": {
                "location": {
                    "type": "string",
                    "description": "The city and state, e.g. San Francisco, CA"
                },
                "unit": {
                    "type": "string",
                    "enum": ["celsius", "fahrenheit"],
                    "description": "The unit of temperature, either \"celsius\" or \"fahrenheit\""
                }
            },
            "required": ["location"]
        }
    
                 """
                )
            )
```


or builder with expectable factory methods:

```java
            .addTool(
                Tool.builder().name("get_weather")
                .inputSchema(
                    Tool.InputSchema.builder()
                        .addRequiredString("location", "The city and state, e.g. San Francisco, CA")
                        .addEnum("unit", "The unit of temperature, either \"celsius\" or \"fahrenheit\"", Temperature.values())
                        .build()
                ).build()
            )
```

or

```java
            .addTool(
                Tool.builder().name("get_weather")
                .inputSchema(
                    Tool.InputSchema.builder()
                        .addRequiredProperty(ToolProperty.string("location", "The city and state, e.g. San Francisco, CA"))
                        .addProperty(ToolProperty.enum("unit", "The unit of temperature, either \"celsius\" or \"fahrenheit\"", Temperature.values()))
                        .build()
                ).build()
            )
```


or

```java
            .addTool(
                Tool.builder().name("get_weather")
                .inputSchema(
                    Tool.InputSchema.builder()
                        .addProperty(ToolProperty.string("location", "The city and state, e.g. San Francisco, CA", ToolProperty.REQUIRED))
                        .addProperty(ToolProperty.enum("unit", "The unit of temperature, either \"celsius\" or \"fahrenheit\"", Temperature.values()))
                        .build()
                ).build()
            )
```

**Comment by dengpan1** [comment]
Is there a temporary workaround until the SDK doesn't officially support getting results?

**Comment by TomerAberbach** [comment]
> Thanks for response, [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/23299544?s=64&amp;u=c4deebbe13dfcb40cdec7b37fec8e42e10f64787&amp;v=4">@TomerAberbach](https://github.com/TomerAberbach?rgh-link-date=2025-01-09T10%3A53%3A47.000Z).
> 
> I saw that module. Isn't `org.apache.httpcomponents.client5:httpclient5` an http client? It's listed here
> 
> [anthropic-sdk-java/anthropic-java-core/build.gradle.kts](https://github.com/anthropics/anthropic-sdk-java/blob/04dddd52df5569219faca1e0ae31aa968c63fdc7/anthropic-java-core/build.gradle.kts#L16)
> 
> Line 16 in [04dddd5](/anthropics/anthropic-sdk-java/commit/04dddd52df5569219faca1e0ae31aa968c63fdc7)
> 
>  implementation("org.apache.httpcomponents.client5:httpclient5:5.3.1")

@katoquro looks like we use it for `org.apache.hc.client5.http.entity.mime.MultipartEntityBuilder`

**Comment by TomerAberbach** [comment]
> Sorry to bother you, why do I always get 403 when I try to access Claude through the api in Java? But I succeed directly in python? Why does Java show that ANTHROPIC_AUTH_TOKEN is still needed?

@Mubai0628 

Without more information I don't know why it's not working. Please file a new issue with more details if it's still not working

**Comment by TomerAberbach** [comment]
> I am trying to implement custom tool, but I find it difficult to find a correct way, how to build shema.
> 
>             // TODO custom tool
>             .addTool(
>                 Tool.builder().inputSchema(
>                     Tool.InputSchema.builder()
>                         .putAdditionalProperty("my_tool_property", /* ??? */)
>                         .build()
>                 )
>             )
> I would expect direct input:
> 
>              .addTool(
>                  """
>                  {
>         "name": "get_weather",
>         "description": "Get the current weather in a given location",
>         "input_schema": {
>             "type": "object",
>             "properties": {
>                 "location": {
>                     "type": "string",
>                     "description": "The city and state, e.g. San Francisco, CA"
>                 },
>                 "unit": {
>                     "type": "string",
>                     "enum": ["celsius", "fahrenheit"],
>                     "description": "The unit of temperature, either \"celsius\" or \"fahrenheit\""
>                 }
>             },
>             "required": ["location"]
>         }
>     
>                  """
>                 )
>             )
> or builder with expectable factory methods:
> 
>             .addTool(
>                 Tool.builder().name("get_weather")
>                 .inputSchema(
>                     Tool.InputSchema.builder()
>                         .addRequiredString("location", "The city and state, e.g. San Francisco, CA")
>                         .addEnum("unit", "The unit of temperature, either \"celsius\" or \"fahrenheit\"", Temperature.values())
>                         .build()
>                 ).build()
>             )
> or
> 
>             .addTool(
>                 Tool.builder().name("get_weather")
>                 .inputSchema(
>                     Tool.InputSchema.builder()
>                         .addRequiredProperty(ToolProperty.string("location", "The city and state, e.g. San Francisco, CA"))
>                         .addProperty(ToolProperty.enum("unit", "The unit of temperature, either \"celsius\" or \"fahrenheit\"", Temperature.values()))
>                         .build()
>                 ).build()
>             )
> or
> 
>             .addTool(
>                 Tool.builder().name("get_weather")
>                 .inputSchema(
>                     Tool.InputSchema.builder()
>                         .addProperty(ToolProperty.string("location", "The city and state, e.g. San Francisco, CA", ToolProperty.REQUIRED))
>                         .addProperty(ToolProperty.enum("unit", "The unit of temperature, either \"celsius\" or \"fahrenheit\"", Temperature.values()))
>                         .build()
>                 ).build()
>             )

@sgflt you can use `JsonValue` like so:
```java
Tool.InputSchema.builder()
  .type(Tool.InputSchema.Type.OBJECT) // Note: we are working on a change that will likely let you omit this
  .properties(JsonValue.from(
    Map.of(
      "location", Map.of("type", "string", "description", "..."),
      "unit", Map.of(
        "type", "string",
        "enum", List.of("celsius", "fahrenheit"),
        "description", "..."
      )
    )
  ))
```

**Comment by TomerAberbach** [comment]
> Is there a temporary workaround until the SDK doesn't officially support getting results?

@dengpan1 no, there's no workaround. But I suspect we'll have support released this week

**Comment by sgflt** [comment]
> [@sgflt](https://github.com/sgflt) you can use `JsonValue` like so:
> 
> Tool.InputSchema.builder()
>   .type(Tool.InputSchema.Type.OBJECT) // Note: we are working on a change that will likely let you omit this
>   .properties(JsonValue.from(
>     Map.of(
>       "location", Map.of("type", "string", "description", "..."),
>       "unit", Map.of(
>         "type", "string",
>         "enum", List.of("celsius", "fahrenheit"),
>         "description", "..."
>       )
>     )
>   ))

OK, thanks for pointing that possibility. It seems to be sufficient for my use case.

But, I am new to AI APIs and for me they seem to be very low level.
I think I am doing things in a wrong way or there may be API improved to get less steep learning curve. I'm just staring at the code, trying to figure out how to parse the responses.

For example:
```java
 final var response = this.client.messages().create(
        messageBuilder
            .addMessage(
                MessageParam.builder()
                    .role(MessageParam.Role.USER)
                    .content(MessageParam.Content.ofString(body))
                    .build()
            ).build()
    );

    final var stopReason =
        response.stopReason().map(Message.StopReason::value).orElse(Message.StopReason.Value._UNKNOWN);
    if (stopReason == Message.StopReason.Value.TOOL_USE) {
      response.content().stream()
          .filter(ContentBlock::isToolUseBlock)
          .findFirst() // assuming there is only one possible tool use block in turn response
          .map(ContentBlock::asToolUseBlock)
          .ifPresent(toolUseBlock -> {
            messageBuilder.addMessage(response.toParam());
            // TODO tool call
            messageBuilder.addMessage(
                MessageParam.builder()
                    .role(MessageParam.Role.USER)
                    .content(MessageParam.Content.ofContentBlockParams(
                        List.of(
                            ContentBlockParam.ofToolResultBlockParam(
                                ToolResultBlockParam.builder()
                                    .toolUseId(toolUseBlock.id())
                                    .content(ContentBlock.ofTextBlock(toolResponse))
                                    .build()
                            )
                        )
                    ))
                    .build()
            );
          });
    }

    return response;
```


I would wish kind of conversational object that does all this automagically:
```java
    final var conversation = new AnthropicConversation();
    conversation.system("system prompt");
    conversation.addTool(toolDefinition, toolHandler);
    conversation.addTool(...);
    conversation.addTool(...);
    conversation.addTool(...);

    conversation.addMessage(
          MessageParam.builder()
              .role(MessageParam.Role.USER)
              .content(MessageParam.Content.ofString(body))
              .build()
    );

    // send initial prompt
    // invoke tools if AI asked tool_use
    // handles all turns internally, maintains USER/ASSISTANT messages
    // return final response as string
    String finalResponse = conversation.talkToCentralBrainOfMankind();
```

**Comment by TomerAberbach** [comment]
@dengpan1 the results endpoints are now shipped as of v0.1.0-alpha.9

**Comment by TomerAberbach** [comment]
@sgflt just an FYI that I've added some examples that you may find helpful: https://github.com/anthropics/anthropic-sdk-java/tree/main/anthropic-java-example/src/main/java/com/anthropic/example

Let me know if you're still finding the SDK hard to use after looking at them

(also make sure to update to the latest version before looking)

**Comment by TomerAberbach** [comment]
This issue has gotten a little out of hand due to the lack of threading, so we're switching to the Discussions feature.

If you have additional feedback, please file a new issue or comment here: https://github.com/anthropics/anthropic-sdk-java/discussions/95
